paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+                    if re.search('ERROR|WARNING|Warning|duplicate|FATAL|Could', line):
+                        log.warning(
+                            'It appears there was a problem with solvation: check `leap.log`...'
+                        )
+        except:
+            return
+
+    def check_for_leap_log(self):
+        """
+        Check if `leap.log` exists in output_path, and if so, delete so the current run doesn't append.
+
+        """
+        filename = 'leap.log'
+        try:
+            os.remove(self.output_path + filename)
+            log.debug('Deleted existing leap.log file: '+self.output_path + filename)
+        except OSError:
+            pass
+
+    def solvate(self):
+        """
+        Solvate a structure with an exact number of waters or buffer size.
+
+        """
+    
+        # If `buffer_target` is a string ending with 'A', an estimate of the number of waters is generated, otherwise,
+        # the target is returned.
+        self.set_target_waters()
+    
+        if self.add_ions:
+            self.set_additional_ions()
+    
+        # First, a coarse adjustment...
+        # This will run for 50 iterations or until we (a) have more waters than the target and (b) are within ~12 waters
+        # of the target (that can be manually removed).
+        cycle = 0
+        while cycle < self.max_cycles:
+            # Start by not manually removing any water, just adjusting the buffer value to get near the target number of
+            # waters...
+
+            # Find out how many waters for *this* buffer value...
+            waters = self.count_waters()
+            self.wat_added_history.append(waters)
+            self.buffer_val_history.append(self.buffer_value)
+            log.debug("Cycle {:02.0f}   {:8.5f} {:6.0f} ({:6.0f})".format(cycle, self.buffer_value, waters, self.target_waters))
+    
+            # Possible location of a switch to adjust the buffer_value by polynomial
+            # fit approach.
+    
+            # If we've nailed it, break!
+            if waters == self.target_waters:
+                # Run one more time and save files
+                self.write_input(include_saves=True)
+                self.run()
+                return
+            # If we are close, go to fine adjustment...
+            elif waters > self.target_waters and (waters - self.target_waters) < 12:
+                self.remove_waters_manually()
+                self.write_input(include_saves=True)
+                self.run()
+                return
+            # Otherwise, try to keep adjusting the number of waters...
+            else:
+                self.adjust_buffer_value()
+                # Now that we're close, let's re-evaluate how many ions to add, in case the volume has changed a lot.
+                # (This could be slow and run less frequently...)
+                ################# HOW DO WE KNOW THAT ONCE EVERY 10 IS GOOD ENOUGH??????????????????????????!!!!!!!!!!!!!!!!!!
```

Comment:
Just a guess, but it shouldn't matter too much if we estimate again once we get into the manual water removal step, so that will definitely nail it. This is simply to prevent running through the cycles adding 10 ions and then realize once we get to the manual step that the box has shrunk so much, we need to only add 1, which then leaves 9 voids.

Reply:
Now that I think about it ... I think we should probably only add ions after we finish getting the waters correct.  Hmm ... does `addionsrand` overwrite water molecules ... I forget.

Reply:
> I think we should probably only add ions after we finish getting the waters correct.

No, because then it will remove water molecules (I think)!

Reply:
Hmm leaving alone for now.